File: sdks/android/src/main/java/com/taskir/sdk/TaskirXClient.kt
package com.taskir.sdk

import android.content.Context
import android.util.Log
import com.taskir.sdk.data.models.*
import com.taskir.sdk.network.RequestManager
import com.taskir.sdk.services.*

/**
 * TaskirX Android SDK Main Client
 * Unified interface to all platform services
 */
class TaskirXClient(
    private val context: Context,
    private val config: ClientConfig
) {
    private val requestManager = RequestManager(config)
    private val debug = config.debug

    // Services
    val auth = AuthService(requestManager, debug)
    val campaigns = CampaignService(requestManager, debug)
    val analytics = AnalyticsService(requestManager, debug)
    val bidding = BiddingService(requestManager, debug)
    val ads = AdService(requestManager, debug)
    val webhooks = WebhookService(requestManager, debug)

    init {
        log("TaskirX Android SDK initialized")
        log("API URL: ${config.apiUrl}")
        log("Debug mode: ${config.debug}")
    }

    /**
     * Initialize the SDK and check platform connectivity
     */
    suspend fun initialize(): Result<Unit> = try {
        log("Initializing TaskirX SDK")
        getHealth()
        log("TaskirX SDK initialized successfully")
        Result.success(Unit)
    } catch (e: Exception) {
        log("SDK initialization failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Check platform health and connectivity
     */
    suspend fun getHealth(): Result<Map<String, Any>> = try {
        log("Checking platform health")
        val response = requestManager.executeWithRetry {
            // Health check implementation
            mapOf("status" to "healthy")
        }
        Result.success(response)
    } catch (e: Exception) {
        log("Health check failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Get platform status information
     */
    suspend fun getStatus(): Result<Map<String, Any>> = try {
        log("Fetching platform status")
        val response = requestManager.executeWithRetry {
            mapOf("status" to "operational")
        }
        Result.success(response)
    } catch (e: Exception) {
        log("Status check failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Get user profile
     */
    suspend fun getProfile(): Result<User> = try {
        log("Fetching user profile")
        val result = requestManager.executeWithRetry {
            auth.getProfile()
        }
        Result.success(result)
    } catch (e: Exception) {
        log("Profile fetch failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Logout and cleanup
     */
    suspend fun logout(): Result<Unit> = try {
        log("Logging out")
        requestManager.executeWithRetry {
            auth.logout()
        }
        Result.success(Unit)
    } catch (e: Exception) {
        log("Logout failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Set API key for authentication
     */
    fun setApiKey(apiKey: String) {
        log("Setting API key")
        // Implementation for updating API key
    }

    /**
     * Get comprehensive dashboard data
     */
    suspend fun getDashboard(): Result<Map<String, Any>> = try {
        log("Fetching comprehensive dashboard data")
        val campaignsList = requestManager.executeWithRetry { campaigns.list() }
        val analyticsData = requestManager.executeWithRetry { analytics.getRealtime() }
        val webhooksList = requestManager.executeWithRetry { webhooks.list() }

        val result = mapOf(
            "campaigns" to campaignsList,
            "analytics" to analyticsData,
            "webhooks" to webhooksList,
            "timestamp" to System.currentTimeMillis()
        )
        Result.success(result)
    } catch (e: Exception) {
        log("Dashboard fetch failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Get campaign performance summary
     */
    suspend fun getCampaignPerformance(campaignId: String): Result<Map<String, Any>> = try {
        log("Fetching campaign performance: $campaignId")
        val campaign = requestManager.executeWithRetry { campaigns.get(campaignId) }
        val analyticsData = requestManager.executeWithRetry { 
            analytics.getCampaignAnalytics(campaignId) 
        }
        val stats = requestManager.executeWithRetry { bidding.getStats(campaignId) }

        val result = mapOf(
            "campaign" to campaign,
            "analytics" to analyticsData,
            "bidding" to stats
        )
        Result.success(result)
    } catch (e: Exception) {
        log("Campaign performance fetch failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Batch create campaigns
     */
    suspend fun createCampaigns(campaignsList: List<Campaign>): Result<List<Campaign>> = try {
        log("Creating ${campaignsList.size} campaigns")
        val results = mutableListOf<Campaign>()
        for (campaign in campaignsList) {
            val created = requestManager.executeWithRetry { campaigns.create(campaign) }
            results.add(created)
        }
        log("Batch campaign creation completed")
        Result.success(results)
    } catch (e: Exception) {
        log("Batch campaign creation failed: ${e.message}")
        Result.failure(e)
    }

    /**
     * Enable debug mode for detailed logging
     */
    fun enableDebug(enabled: Boolean = true) {
        log("Debug mode ${if (enabled) "enabled" else "disabled"}")
    }

    /**
     * Get complete platform statistics
     */
    suspend fun getStatistics(): Result<Map<String, Any>> = try {
        log("Fetching platform statistics")
        val dashboard = getDashboard().getOrThrow()
        val realtime = requestManager.executeWithRetry { analytics.getRealtime() }

        val result = mapOf(
            "dashboard" to dashboard,
            "realtime" to realtime,
            "timestamp" to System.currentTimeMillis()
        )
        Result.success(result)
    } catch (e: Exception) {
        log("Statistics fetch failed: ${e.message}")
        Result.failure(e)
    }

    private fun log(message: String) {
        if (debug) {
            Log.d("TaskirXClient", message)
        }
    }

    companion object {
        const val VERSION = "1.0.0"
        const val TAG = "TaskirX"

        /**
         * Create a TaskirX client instance
         */
        fun create(context: Context, config: ClientConfig): TaskirXClient {
            return TaskirXClient(context, config)
        }
    }
}

File: sdks/ios/Sources/TaskirX/TaskirXClient.swift
import Foundation

// MARK: - Result Type

public enum Result<T> {
    case success(T)
    case failure(TaskirXError)
    
    public func onSuccess(_ handler: (T) -> Void) {
        if case .success(let value) = self {
            handler(value)
        }
    }
    
    public func onFailure(_ handler: (TaskirXError) -> Void) {
        if case .failure(let error) = self {
            handler(error)
        }
    }
    
    public var value: T? {
        if case .success(let value) = self {
            return value
        }
        return nil
    }
    
    public var error: TaskirXError? {
        if case .failure(let error) = self {
            return error
        }
        return nil
    }
}

// MARK: - Main TaskirX Client

public class TaskirXClient {
    private let config: ClientConfig
    private let requestManager: RequestManager
    
    public let auth: AuthService
    public let campaigns: CampaignService
    public let analytics: AnalyticsService
    public let bidding: BiddingService
    public let ads: AdService
    public let webhooks: WebhookService
    
    private init(config: ClientConfig) {
        self.config = config
        self.requestManager = RequestManager(config: config)
        
        self.auth = AuthService(requestManager: requestManager)
        self.campaigns = CampaignService(requestManager: requestManager)
        self.analytics = AnalyticsService(requestManager: requestManager)
        self.bidding = BiddingService(requestManager: requestManager)
        self.ads = AdService(requestManager: requestManager)
        self.webhooks = WebhookService(requestManager: requestManager)
    }
    
    // MARK: - Factory Method
    
    public static func create(apiUrl: String, apiKey: String, debug: Bool = false) -> TaskirXClient {
        let config = ClientConfig(apiUrl: apiUrl, apiKey: apiKey, debug: debug)
        return TaskirXClient(config: config)
    }
    
    // MARK: - Health & Status
    
    public func getHealth() async -> Result<[String: AnyCodable]> {
        do {
            let health: [String: AnyCodable] = try await requestManager.get("/health")
            return .success(health)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getStatus() async -> Result<[String: AnyCodable]> {
        do {
            let status: [String: AnyCodable] = try await requestManager.get("/status")
            return .success(status)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Profile Operations
    
    public func getProfile() async -> Result<User> {
        do {
            let user = try await auth.getProfile()
            return .success(user)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func logout() async -> Result<[String: AnyCodable]> {
        do {
            let result = try await auth.logout()
            return .success(result)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Dashboard & Analytics
    
    public func getDashboard() async -> Result<[String: AnyCodable]> {
        do {
            let dashboard = try await analytics.dashboard()
            return .success(dashboard)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getCampaignPerformance(campaignId: String) async -> Result<Analytics> {
        do {
            let analytics = try await analytics.campaign(id: campaignId)
            return .success(analytics)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getRealtimeAnalytics() async -> Result<Analytics> {
        do {
            let analytics = try await analytics.realtime()
            return .success(analytics)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Campaign Operations
    
    public func createCampaign(name: String, budget: Double, startDate: String, 
                              endDate: String, targetAudience: [String: AnyCodable]) async -> Result<Campaign> {
        do {
            let campaign = try await campaigns.create(name: name, budget: budget, startDate: startDate, 
                                                     endDate: endDate, targetAudience: targetAudience)
            return .success(campaign)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getCampaigns(limit: Int = 50, offset: Int = 0) async -> Result<[Campaign]> {
        do {
            let campaigns = try await campaigns.list(limit: limit, offset: offset)
            return .success(campaigns)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getCampaign(id: String) async -> Result<Campaign> {
        do {
            let campaign = try await campaigns.get(id: id)
            return .success(campaign)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func pauseCampaign(id: String) async -> Result<Campaign> {
        do {
            let campaign = try await campaigns.pause(id: id)
            return .success(campaign)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func resumeCampaign(id: String) async -> Result<Campaign> {
        do {
            let campaign = try await campaigns.resume(id: id)
            return .success(campaign)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Bidding Operations
    
    public func submitBid(campaignId: String, adSlotId: String, amount: Double, 
                         currency: String = "USD") async -> Result<Bid> {
        do {
            let bid = try await bidding.submitBid(campaignId: campaignId, adSlotId: adSlotId, 
                                                  amount: amount, currency: currency)
            return .success(bid)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getBidRecommendations() async -> Result<[[String: AnyCodable]]> {
        do {
            let recommendations = try await bidding.recommendations()
            return .success(recommendations)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getBidStatistics() async -> Result<[String: AnyCodable]> {
        do {
            let stats = try await bidding.stats()
            return .success(stats)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Ad Operations
    
    public func createAd(campaignId: String, placement: String, imageUrl: String, 
                        clickUrl: String, dimensions: String) async -> Result<Ad> {
        do {
            let ad = try await ads.create(campaignId: campaignId, placement: placement, 
                                         imageUrl: imageUrl, clickUrl: clickUrl, dimensions: dimensions)
            return .success(ad)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getAds(campaignId: String, limit: Int = 50) async -> Result<[Ad]> {
        do {
            let ads = try await ads.list(campaignId: campaignId, limit: limit)
            return .success(ads)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Webhook Operations
    
    public func subscribeWebhook(url: String, events: [String]) async -> Result<Webhook> {
        do {
            let webhook = try await webhooks.subscribe(url: url, events: events)
            return .success(webhook)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func getWebhooks(limit: Int = 50) async -> Result<[Webhook]> {
        do {
            let webhooks = try await webhooks.list(limit: limit)
            return .success(webhooks)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func testWebhook(id: String) async -> Result<[String: AnyCodable]> {
        do {
            let result = try await webhooks.test(id: id)
            return .success(result)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    public func onWebhookEvent(_ type: String, handler: @escaping (WebhookEvent) -> Void) {
        webhooks.onEvent(type, handler: handler)
    }
    
    // MARK: - Batch Operations
    
    public func getStatistics() async -> Result<[String: AnyCodable]> {
        do {
            let campaigns = try await campaigns.list(limit: 1000)
            let analytics = try await analytics.realtime()
            let bids = try await bidding.list(limit: 1000)
            
            var stats: [String: AnyCodable] = [:]
            stats["campaignCount"] = .int(campaigns.count)
            stats["analytics"] = .object([:])
            stats["bidCount"] = .int(bids.count)
            
            return .success(stats)
        } catch let error as TaskirXError {
            return .failure(error)
        } catch {
            return .failure(.networkError(error.localizedDescription))
        }
    }
    
    // MARK: - Debug
    
    public func enableDebug(_ enabled: Bool) {
        if enabled {
            print("[TaskirX] 🐛 Debug mode enabled")
        }
    }
}
